fixes #15919 statistics page is now loaded via AJAX

ohadlevy opened the pull request.

ohadlevy commented:
test failures seems unrelated.

ohadlevy commented:
@tbrisker fixed rubocop.

ohadlevy commented:
rebased

ohadlevy commented:
[test]

sean797 commented:
I'm not sure how this relates to theforeman/foreman#3438 or if it's worth thinking about?

ohadlevy commented:
@sean797 - nice - I missed that PR.

I could see how the two can integrate, but in that case I would also like to discuss dashboard widgets (e.g. what is the different between a widget and statistics model?)

I believe my patch will simplify your patch (e.g. in terms of statistics chart type - host counter, fact counter etc).

ohadlevy commented:
tests are passing, its just katello job which is unstable atm.

tbrisker commented:
Seems like somewhere some permissions are missing, when trying to view as non-admin this is what it looks like:
![screenshot from 2016-08-03 11-45-29](https://cloud.githubusercontent.com/assets/1540531/17359653/448c7710-5970-11e6-813e-10ff6c3cc0b2.png)

tbrisker commented on app/views/statistics/index.html.erb:
```diff
@@ -1,6 +1,10 @@
+  <% @stats.each do |stat| %>
+    <div data-ajax-url=<%= stat.url %> class='stats-well' data-on-complete='refreshCharts'>
+      <div class='statistics-pie-placeholder'>
+        <h4 class='ca pie-title'> <span class='spinner spinner-xl spinner-inline'> </span> <%= stat.title %> </h4>
```
There is no `spinner-xl`, the largest is `spinner-lg`. Also, looks strange to have the spinner inline before the title, maybe move it down to where the chart should be?

ohadlevy replied:
I'll update to lg, but lg by itself looks bad on such a large space, do you
happen to have a better css alternative for me? :)

thanks

tbrisker commented on app/helpers/statistics_helper.rb:
```diff
@@ -1,15 +1,6 @@
-      flot_pie_chart("swap_usage",_("Average swap usage"), [{:label=>_("free swap"), :data=>@swap_free},{:label=>_("used swap"), :data=>@swap_size-@swap_free}], options)
-    ]
+  def chart(stat)
+    options = {:class => 'statistics-pie small', :expandable => true, :'border' => 0, :show_title => true}
```
nitpick: why `:'border'` and not `:border`?

tbrisker commented on app/services/statistics.rb:
```diff
@@ -0,0 +1,14 @@
+      CountHosts.new(:count_by => :operatingsystem, :title => ("OS Distribution"), :search =>"os_title=~VAL~"),
+      CountHosts.new(:count_by => :architecture, :title => _("Architecture Distribution"), :search => "facts.architecture=~VAL~"),
+      CountHosts.new(:count_by => :environment, :title => _("Environment Distribution"), :search => "environment=~VAL~"),
+      CountFacts.new(:count_by => 'processorcount', :unit => Nn_('%s core', '%s cores'), :title => "Number of CPUs", :search => "facts.processorcount=~VAL1~"),
```
why are some count_by strings and some symbols?

tbrisker commented on app/assets/stylesheets/charts.scss:
```diff
@@ -62,6 +62,12 @@
+  height: 268px;
+}
+
 .statistics-pie{
```
instead of duplicating the css above, you can change this to `.statistics-pie, .statistics-pie-placeholder{` and add the `small` class to the placeholder div.

ohadlevy replied:
@tbrisker there is some jquery hook that uses that class, it was my first attempt but it broke something else.

tbrisker replied:
yes, `.statistics-pie` is used for the charts initialization, which is why I suggested adding the placeholder class to use the same layout (see comment below on the html)

ohadlevy replied:
fixed

tbrisker commented on app/views/statistics/index.html.erb:
```diff
@@ -1,6 +1,10 @@
+    <div data-ajax-url=<%= stat.url %> class='stats-well' data-on-complete='refreshCharts'>
+      <div class='statistics-pie-placeholder'>
+        <h4 class='ca pie-title'> <span class='spinner spinner-xl spinner-inline'> </span> <%= stat.title %> </h4>
+      </div>
```
To get the similar layout as the chart, use something like:

```
    <div data-ajax-url=<%= stat.url %> class='stats-well' data-on-complete='refreshCharts'>
      <h4 class='ca pie-title'><%= stat.title %></h4>
      <div class='statistics-pie-placeholder small'>
        <span class='spinner spinner-lg spinner-inline'></span>
      </div>
    </div>
```

tbrisker commented:
@ohadlevy Thanks, looks good in general, just a few minor comments and one permissions issue.

ohadlevy commented:
@tbrisker fixed all of your comments beside the spinner location, looking for suggestions on this one.

ohadlevy commented:
@tbrisker I can't reproduce your issue with permissions, can you provide a trace?

ohadlevy committed:
fixes #15919 - statistics page is loaded via AJAX

Also:
- refactor how statistics data is being called, now it should be trivial
to add additional charts or consume other data from it. It should make
it much easier to reuse statistics charts in other places
(e.g. dashboard) or unify how fact charts are done.
- remove unused JSON response from the statistics controller.

tbrisker commented on app/services/statistics/base.rb:
```diff
@@ -0,0 +1,29 @@
+    end
+
+    def id
+      @id || count_by.to_s
```
why not move this into the initializer? also, why not leave it as a symbol?

ohadlevy replied:
some of the queries required string (e.g. facts) instead of changing that and everywhere that uses it, I chose to change it here.

tbrisker commented on app/services/statistics/base.rb:
```diff
@@ -0,0 +1,29 @@
+    end
+
+    def url
+      @url || "statistics/#{id}"
```
same, why not move this to the initializer?

ohadlevy replied:
I think it make more sense to override it in child class, i find it more readable.

tbrisker commented on app/services/statistics/base.rb:
```diff
@@ -0,0 +1,29 @@
+      @url || "statistics/#{id}"
+    end
+
+    def metadata
```
where is this used except in the test?

ohadlevy replied:
its not, it was based on a discussion with Gail, we can remove it and use
it later on when we change the response to pure JS.

ohadlevy replied:
but I would prefer to keep it.

tbrisker commented:
merged as 4202625d3c2bfc4cf90f792c4246029743163c45 with minor changes, thanks @ohadlevy!
